kotlinx-coroutines-core/common/test/channels/ClosedChannelTest.kt

```diff
@@ -0,0 +1,196 @@
+package kotlinx.coroutines.channels
+
+import kotlinx.coroutines.testing.*
+import kotlin.test.*
+
+class ClosedChannelTest : TestBase() {
+    /**
+     * Iterator.
+     */
+
+    @Test
+    fun testIteratorClosedHasNextFalse() = runTest {
+        TestChannelKind.entries.forEach { kind ->
+            val channel = kind.create<Int>()
+            val iterator = channel.iterator()
+            channel.close()
+            assertFalse(iterator.hasNext())
+        }
+    }
+
+    @Test
+    fun testIteratorClosedWithExceptionHasNextThrows() = runTest {
+        TestChannelKind.entries.forEach { kind ->
+            val channel = kind.create<Int>()
+            val iterator = channel.iterator()
+            channel.close(TestException())
+            assertFailsWith<TestException> { (iterator.hasNext()) }
+        }
+    }
+
+    @Test
+    fun testClosedToListOk() = runTest {
+        TestChannelKind.entries.forEach { kind ->
+            val channel = kind.create<Int>()
+            channel.close()
+            channel.toList()
+        }
+    }
+
+    @Test
+    fun testClosedWithExceptionToListThrows() = runTest {
+        TestChannelKind.entries.forEach { kind ->
+            val channel = kind.create<Int>()
+            channel.close(TestException())
+            assertFailsWith<TestException> { channel.toList() }
+        }
+    }
+
+    @Test
+    fun testClosedConsumeEachOk() = runTest {
+        TestChannelKind.entries.forEach { kind ->
+            val channel = kind.create<Int>()
+            channel.close()
+            channel.consumeEach {  }
+        }
+    }
+
+    @Test
+    fun testClosedWithExceptionConsumeEachThrows() = runTest {
+        TestChannelKind.entries.forEach { kind ->
+            val channel = kind.create<Int>()
+            channel.close(TestException())
+            assertFailsWith<TestException> { channel.consumeEach {  } }
+        }
+    }
+
+    /**
+     * toList() alternative when a channel might close with an exception.
+     */
+
+    @Test
+    fun testToListCatching() = runTest {
+        TestChannelKind.entries.forEach { kind ->
+            val channel = kind.create<Unit>()
+            channel.close(TestException())
+            val lst = mutableListOf<Unit>()
+            while (true) {
```

Comment:
Why is this loop necessary? Only one iteration of it will run. `lst:add` will never run, as won't `getOrThrow()`. The whole test can be simplified to
```kotlin
            channel.close(TestException())
            val result = channel.receiveCatching()
            result.onClosed { // suspends
                assertTrue(it is TestException)
            }
            assertTrue(result.isClosed)
```
This is covered at least by `BasicOperationsTest#testReceiveCatching`